src/main/java/spoon/reflect/declaration/CtFormalTypeDeclarer.java

```diff
@@ -43,4 +45,9 @@
 	 * Removes a type parameters from this generic element.
 	 */
 	boolean removeFormalCtTypeParameter(CtTypeParameter formalTypeParameter);
+
+	/**
+	 * @return the instance of {@link GenericTypeAdapter} which can adapt type parameters from different scopes to scope of this declarer
+	 */
+	GenericTypeAdapter getGenericTypeAdapter();
```

Comment:
I use it here:
```java
class CtTypeParameterImpl {
...
	public boolean isSubtypeOf(CtTypeReference<?> superTypeRef) {
		if (superTypeRef instanceof CtTypeParameterReference) {
			//the type is type parameter too. Use appropriate sub type checking algorithm
			CtTypeParameter superTypeParam = (CtTypeParameter) superTypeRef.getDeclaration();
			return isSubtypeOf(getTypeParameterDeclarer().getGenericTypeAdapter(), this, superTypeParam);
		}
		//type is normal type
		return getTypeErasure().isSubtypeOf(superTypeRef);
	}
```
Without that `getGenericTypeAdapter` method, I would need a `if` with 3 cases:
1) CtType
2) CtMethod
3) CtConstructor
where each case calls appropriate constructor of GenericTypeAdapter implementation.

So this method actually represents a way how to create instance of implementation of GenericTypeAdapter. 

What about adding a spoon factory method instead? Please suggest name and factory class if you agree. Should I use an AbstractSpoonVisitor instead of ugly `if`?

I am waiting for your feedback before I (re)move that code.